### ISSUE TITLE
TPU auto_optim barrier is always True

### ISSUE THREAD

**Issue by ryanwongsa** [pytorch/ignite#1535]
## 🐛 Bug description

In the [auto_optim](https://github.com/pytorch/ignite/blob/cfea1e4c9f4d4ed7767b3b8390a5ed6492a001e7/ignite/distributed/auto.py#L211) method when using TPUs should the `barrier=True` always be set to True when overriding the [step](https://github.com/pytorch/ignite/blob/cfea1e4c9f4d4ed7767b3b8390a5ed6492a001e7/ignite/distributed/auto.py#L330). 
Unless I am understanding it incorrectly it should be changed depending on the number of TPU devices. Looking through the [pytorch 1.5 docs](https://pytorch.org/xla/release/1.5/index.html#running-on-multiple-xla-devices-with-multiprocessing) it should only be set to `True` when using a single xla device. 

*EDIT*: It looks like something changed from 1.5 to 1.7 with how xla works. [pytorch 1.7](https://pytorch.org/xla/release/1.7/index.html#running-on-multiple-xla-devices-with-multiprocessing) doesn't use barrier but `xm.mark_step()` for single xla devices

## Environment

 - PyTorch Version (e.g., 1.4):
 - Ignite Version (e.g., 0.3.0):
 - OS (e.g., Linux):
 - How you installed Ignite (`conda`, `pip`, source):
 - Python version:
 - Any other relevant information:

**Comment by vfdev-5** [comment]
@ryanwongsa thanks for reporting ! Yes, I agree that we should configure `xm.mark_step()` usage depending on the number of XLA devices used.
If you would like to send a PR with a fix, it would be very helpful 👍 
We can think about using [this test](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py) as reference for our code. And maybe put `xm.mark_step()` as an option if user calls something like `xm.add_step_closure` in the training loop...

**Comment by ryanwongsa** [comment]
I was looking into this for a bit as I noticed my training process would hang after a while when using TPU. It turns out the issue was not related to the `barrier` issue but related to wandb model watching issue.  I tried changing the auto_optim approach to set `barrier=False` with something like below:

```python
    class _XLADistributedOptimizer(Optimizer):
        def __init__(self, optimizer):
            super(self.__class__, self).__init__(optimizer.param_groups)
            self.wrapped_optimizer = optimizer

        def step(self, closure=None):
            if idist.get_nproc_per_node()>1:
                barrier = False
            else:
                barrier = True
            xm.optimizer_step(self.wrapped_optimizer, barrier=barrier)
```

and there doesn't seem to be any noticeable change. I am only able to test with python 1.5 as Kaggle notebooks with TPUs seem to install pytorch 1.5 when using TPUs.  I am unable to test with Python 1.7 using `xm.mark_step()` so not sure if it may be an issue when Kaggle switches to python 1.7. My knowledge of how TPUs is limited so I might be misunderstanding what affect the `barrer` parameter has on training.

TL;DR:  The issue initially raised might not actually be an issue.

**Comment by vfdev-5** [comment]
@ryanwongsa I see. The purpose of `barrier` (as far as I understand it) is to execute current graph. If you see their code, `barrier=True` means to call `mark_step` behind:

https://github.com/pytorch/xla/blob/a2c0db47356a532d64b4445d9372b1291dfbbc9e/torch_xla/core/xla_model.py#L783

and if we are using `MpDeviceLoader/ParallelLoader`, `mark_step` is already called on `next()`:

https://github.com/pytorch/xla/blob/3eaee46ef679cc6a0f1f694bd0a007dbfd09c51b/torch_xla/distributed/parallel_loader.py#L37

Thus, if we keep call it twice, probably this could lead to possibly unsignificant slowdown (it would be nice to measure it). Probably, we can add a `**kwargs` to `auto_optim` : 
```python
def auto_optim(optimizer: Optimizer, **kwargs: Any) -> Optimizer:

    bnd = idist.backend()
    if idist.has_xla_support and bnd == idist_xla.XLA_TPU:
        cls = type(optimizer.__class__.__name__, (optimizer.__class__,), dict(_XLADistributedOptimizer.__dict__))
        return cls(optimizer, **kwargs)

    if idist.has_hvd_support and bnd == idist_hvd.HOROVOD:
        import horovod.torch as hvd

        optimizer = hvd.DistributedOptimizer(optimizer, **kwargs)
        hvd.broadcast_optimizer_state(optimizer, root_rank=0)
        return optimizer

    return optimizer


    class _XLADistributedOptimizer(Optimizer):
        def __init__(self, optimizer: Optimizer, **kwargs: Any) -> None:
            super(self.__class__, self).__init__(optimizer.param_groups)  # type: ignore[call-arg]
            self.wrapped_optimizer = optimizer
            self.barrier = kwargs.get("barrier", False) 

        def step(self, closure: Optional[Callable] = None) -> None:
            xm.optimizer_step(self.wrapped_optimizer, barrier=self.barrier)
```
such that we can control this. What do you think ?

>  I am unable to test with Python 1.7 using xm.mark_step() so not sure if it may be an issue when Kaggle switches to python 1.7. 

Probably, we can install it manually like:
```
!curl https://raw.githubusercontent.com/pytorch/xla/master/contrib/scripts/env-setup.py -o pytorch-xla-env-setup.py
!python pytorch-xla-env-setup.py --apt-packages libomp5 libopenblas-dev
```
and update torch too ...

**Comment by ryanwongsa** [comment]
If it is changed to be a parameter in the `auto_optim` method then the user needs to know to do the `xm.mark_step()` if they set `barrier=False` as I could be wrong but I thought the purpose of `auto_optim` and custom `step` was so that it can be generalised without worrying about which device (cpu, gpu, tpu) is being used.

**Comment by vfdev-5** [comment]
@ryanwongsa I agree with you. My idea was to add this option such that we can have a possibility to control that when we surely know that we are using tpus and we still use `auto_*` methods. But, thinking more about your comment, you are right that this option can be misleading for other backends... In this case we can keep barrier=True and accept this unsignificant slowdown for calling barrier twice...

**Comment by ryanwongsa** [comment]
Sure I agree, I am happy to close the issue for now. I will try to test the impact of the performance of `barrier` some time when I have the TPU resources as currently my 30 hours per week in Kaggle notebooks runs out pretty quickly.

**Comment by vfdev-5** [comment]
> I will try to test the impact of the performance of barrier some time when I have the TPU resources as currently my 30 hours per week in Kaggle notebooks runs out pretty quickly.

Thanks ! Maybe, a good start can be on Google Colab and then on Kaggle :)
